File: BinaryTreeFromInorderAndPostorder.py
#Time Complexity - O(n)
#Space Complexity - O(Height of the tree)
# Problem successfully executed on leetcode
# No problems faced while coading this

class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if inorder==None or postorder==None or len(inorder)==0:
            return None
        
        root=TreeNode(postorder[len(postorder)-1])
        mid = inorder.index(root.val)
        root.left=self.buildTree(inorder[:mid],postorder[:mid])
        root.right=self.buildTree(inorder[mid+1:],postorder[mid:len(postorder)-1])
        
        
        return root

File: BinaryTreeFromInorderAndPostOrder.java
//Time: O(N^2) | Space: O(N^2)
// Brute force
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(postorder.length == 0) return null;
        int rootVal = postorder[postorder.length-1];
        TreeNode root = new TreeNode(rootVal);
        int rootInorderIdx = -1;
        for(int i=0;i<inorder.length;i++) {
            if(inorder[i] == rootVal){
                rootInorderIdx = i;
                break;
            }
        }
        int[] inLeft = Arrays.copyOfRange(inorder, 0, rootInorderIdx);
        int[] inRight = Arrays.copyOfRange(inorder, rootInorderIdx+1, inorder.length);
        int[] postLeft = Arrays.copyOfRange(postorder, 0, postorder.length-inRight.length-1);
        int[] postRight = Arrays.copyOfRange(postorder, postorder.length-1-inRight.length, postorder.length-1);
        root.left = buildTree(inLeft, postLeft);
        root.right = buildTree(inRight, postRight);
        return root;
    }
}

// Time:O(N) | Space: O(N)
// Efficient soln using hashMap on inorder
// Traversing the postorder from last, finding its left and subtrees on inorder, adding the root to the result
class Solution {
    Map<Integer, Integer> map = new HashMap<>();
    int postIndex;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(postorder.length == 0) return null;
        postIndex = postorder.length-1;
        for(int i=0;i<inorder.length;i++) {
            map.put(inorder[i], i);
        }
        return builder(postorder, 0, postorder.length-1);
    }
    private TreeNode builder(int[] postorder, int start, int end) {
        if(start > end) return null;
        int rootVal = postorder[postIndex--];
        TreeNode root = new TreeNode(rootVal);
        int rootValIndx = map.get(rootVal);
        root.right = builder(postorder, rootValIndx+1,end);
        root.left = builder(postorder,start, rootValIndx-1);
        return root;
    }
}